README.md
![RestyGWT](http://resty-gwt.github.io/images/restygwt-logo.png)
==============

* [![Build Status](https://secure.travis-ci.org/resty-gwt/resty-gwt.png)](http://travis-ci.org/resty-gwt/resty-gwt)

To know more have a look to [RestyGWT web site](http://resty-gwt.github.io/)

Description
-----------

RestyGWT is a GWT generator for REST services and JSON encoded data transfer objects.

Features
--------

* Generates Async Restful JSON based service proxies
* Java Object to JSON encoding/decoding
* Easy to use REST API

Releasing
---------

mvn release:prepare -Prun-examples
mvn release:perform



changelog.md
# ![RestyGWT](http://restygwt.fusesource.org/images/restygwt-logo.png)

## [resty-gwt 1.4](http://restygwt.fusesource.org/blog/release-1-4.html), 2013-12-17

* Support enum subtypes by serializing enums implementing an interface with a complex object containing the class name and the value. Issue #164 + Make sure we look up
* Added a test exposing a bug in the generated serializers for enums implementing interfaces with @JsonSubTypes. Issue #164
* Added test and solution for jsonsubtypes with interfaces, does not work with enums yet. Issue #164
* Added support for decoding long strings as proper long values in restygwt, added a testcase verifying the behaviour. Long numbers (not strings) will remain an issue 
* Fixed: Cannot compile generated JsonEncoderDecoder with certain type hierarchy
* Fixed regression where field type is more specific than constructor argument, issue #162.
* getField can throw assertion error if you ask for super types, use the field matched instead for fields from super types
* Set needNullHandling to false only in case of custom generator and true otherwise
* Added Custom Serializer Generators
* RestyGWT now supports normal inheritance, subtypes can be listed recursively.
* cherry picked commit 569b812 of PR #149 and adjusted mock-testcase
* fix #156
* fix naming for inner classes #158
* get more test working or clearly comment out tests in suite
* use rails3-maven-plugin which works with maven-3.1.x as well
* allow Attribute annotation expression to return null value
* getters and setters do not need a matching field. @JsonIgnore works on fields or the setter or the getter
* encode/decode nested arrays, lists, sets and maps as fields or property of a pojo
* move all GWT.lo to java.util.logging.Logger
* send enum null values as null when defined as pojo field #154
* adjusted contribution part and issue tracker to be in line with how we are doing it
* make the URLencode working with null values as well
* reduce noise during test run
* white spaces
* url encode PathParam -  issue #147
* Fix typos, improve consistency
* some white space changes
* Refactor the getPossibleType method in JsonEncoderDecoderClassCreator. I created a kind of JsonTypeInfo.Id visitor that can be reused for other matters. I extracted the code from the JsonEncoderDecoderClassCreator to put it in the visitor.
* Split the HUGE generate method into different methods.
* Fix Eclipse web.xml validation
* Support default value for @JsonTypeInfo property
* Revert "Default @JsonTypeInfo Id.CLASS property value"
* Revert "Add support for JsonDeserialize for interfaces"
* Add support for JsonDeserialize for interfaces
* Default @JsonTypeInfo Id.CLASS property value
* fixed mapping for @JsonTypeInfo - MINIMAL_CLASS (. as prefix); added support for discovering all subtypes of generic type during the generation process, thus no need to use the @JsonSubTypes annotation in combination with @JsonTypeInfo - CLASS or MINIMAL_CLASS
* added a note on hwo to parse jsonString and about JsonProperty to customize names
* added test case for decoding/encoding nested Maps
* tests for methods returning Request or JsonpRequest
* let RestService interface methods return either JsonpRequest or Request
* Fixed 204 not handled roperly for MSIE 8 and 9
* allow getters and setters corresponding field
* added generic list encoder - start to nest collection
* Reduced loglevel for type encoder information to debug
* Fixed a bug in serializing lists and arrays using @FormParams.
* some jsonp test
* make the resource creation lazy inside restService impl - #114
* Support FormParams. Fixes issue #92.
* fix encoding of BigDecimal keys of maps (#111)
* fixed issue with short decoding, cast requred (#106)
* parse strings of generic map keys correctly
* let jsonp handle list of POJOs correctly
* let rest-service use default dispatcher from Defaults when none is set on interface
* Take dispatcher from Defaults if local one for the Method is not set.
* get the generated ode compile
* added missing annotation
* discriminate Lists on JSONP results
* Revert "discriminate JSONArray on JSONP results"
* discriminate JSONArray on JSONP results
* fixed #108 - allow to ignore any fields in the classhierachy
* fixed map decode with complex and number keys
* Documented DirectRestService usage in the user documentation.
* added test for QueryParam
* add test case for @PathParams
* remove yellow flag from eclipse
* fix regression issue #102 : but expression insde brackets to allow further arithemitek to keep expression as is.

## [resty-gwt 1.3](http://restygwt.fusesource.org/blog/release-1-3.html), 2012-08-23

* Added additional tests to ensure ObjectEncoderDecoder is not left out again
* fix  Issue #84: allow null values in collections of pojo and primitives
* XSRFTokenCallbackFilter didn't set the XSRF token correctly in IE9
* allow jsonignore to omit fields for de/serialization - issue #30
* added test for arrays de/encoding
* added support for one-dimensional arrays
* go through ALL callbacks even in case of error
* restful caching shall not cache collection
* fix for #31
* redone reverted patch from StephanBeutel against latest master branch
* new invoker plugin version
* use gwt 2.4.0
* added testcase for passing on the headers when getting GET-method, etc
* remove debug System.out.println
* added testcase for issue #84
* move easymock tests to mokcing package
* remove obsolete files
* since implementation changed we need an extra call
* remove unwanted artifact prefix from previous commit
* obey failed status codes as without cache.
* use getter/setter for token to allow subclassing
* allow null values for primitives and fill use the default values in such cases
* remove yellow flags in eclipse, i.e. cleanup imports, deprectated method, suppress warnings
* remove error in eclipse indigo complaining about not supported goal ... from pullrequest #85
* Add support for numbers and booleans to be decoded to strings

restygwt-website/ext/Website.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import org.fusesource.scalate.RenderContext

package

/**
 * <p>
 * </p>
 *
 * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
 */
object Website {

  val project_name= "RestyGWT"
  val project_slogan= "Making Restful Services Accessible to GWT Applications"
  val project_id= "restygwt"
  val project_jira_key= "RESTYGWT"
  val project_issue_url= "http://github.com/chirino/resty-gwt/issues"
  val project_forums_url= "http://groups.google.com/group/restygwt"
  val project_wiki_url= "http://github.com/chirino/resty-gwt/"
  val project_logo= "/images/restygwt-logo.png"
  val project_version= "1.4"
  val project_snapshot_version= "2.0-SNAPSHOT"

  val project_versions = List(
        project_version,
        "1.3",
        "1.2",
        "1.1",
        "1.0")  


  val project_keywords= "rest,gwt,restygwt,java,javascript,json,xml,dto"

  // -------------------------------------------------------------------
  val github_page = "http://github.com/chirino/resty-gwt"
  val git_user_url = "git://github.com/chirino/resty-gwt.git"
  val git_commiter_url = "git@github.com:chirino/resty-gwt.git"
  
  val project_maven_groupId= "org.fusesource.restygwt"
  val project_maven_artifactId= "restygwt"

  val website_base_url= "http://restygwt.fusesource.org"
}
